pkg/workspace/assert.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package workspace
+
+import (
+	"fmt"
+	"github.com/k14s/starlark-go/starlark"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/filepos"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/template"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamlmeta"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamltemplate"
+)
+
+// Declare @assert/... annotation names
+const (
+	AnnotationAssertValidate template.AnnotationName = "assert/validate"
+)
+
+// ProcessAndRunValidations takes a root Node and traverses the tree checking for assert annotations.
+// Validations are processed and executed using the value of the annotated node as the parameter to the assertions.
+//
+// When the assertions have violations, the errors are collected and stored in the checker.
+// Otherwise, returns nil.
+func ProcessAndRunValidations(n yamlmeta.Node, checker yamlmeta.Visitor) error {
+	if n == nil {
+		return nil
+	}
+
+	err := yamlmeta.Walk(n, checker)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type assertChecker struct {
+	thread     *starlark.Thread
+	violations []error
+}
+
+func newAssertChecker(threadName string, _ *TemplateLoader) *assertChecker {
+	return &assertChecker{thread: &starlark.Thread{Name: threadName}, violations: []error{}}
+}
```

Comment:
If I was tracking correctly, we talked about the `TemplateLoader` param being forward-looking... that we'll likely need it at some point?

I think we should defer anything like this until we actually see the need arise.

I believe that there's even reason to think that it may not be needed for invoking a function: Starlark function values contain (all?) of their execution context and likely won't be loading other templates.

```suggestion
func newAssertChecker(threadName string) *assertChecker {
	return &assertChecker{thread: &starlark.Thread{Name: threadName}, violations: []error{}}
}
```